### PR TITLE
Fix boolean cast in Fortran code

### LINKED ISSUE #1785: Missing boolean cast in Fortran code
### Describe the bug
 Python scripts written to test the fix for the issue #1762 #1641, results in compilation error for the fortran language. Please let me know if there is any existing issue which tracks this behavior. 

### To Reproduce
Provide code to reproduce the behavior:
Example 1:
```python
@template('T1', ['bool[:,:]', 'int[:,:]', 'int8[:,:]', 'int16[:,:]', 'int32[:,:]', 'int64[:,:]', 'float[:,:]', 'float32[:,:]', 'float64[:,:]', 'complex64[:,:]', 'complex128[:,:]'])
def create_bool_copy(a : 'T1'):
    c = np.array(a, dtype='bool')
    return c

if __name__ == '__main__':
    arr = np.ones((2,2), dtype='int16')
    arr[1, 1] = 256
    arr[0, 1] = -1
    arr[1, 0] = 64
    arr[0, 0] = -1
    print(arr)
    arr2 = create_bool_copy(arr)
    print(arr2)
```
Example 2:
```python
def create_float_copy(a : 'bool[:,:]'):
    c = np.array(a, dtype='float32')
    return c

if __name__ == '__main__':
    arr = np.ones((2,2), dtype='bool')
    print(arr)
    arr2 = create_float_copy(arr)
    print(arr2)
```                                                                                              
### Error details
Provide the generated code, or the error message:
```
ERROR at Fortran compilation stage
Traceback (most recent call last):
  File "/home/varadarajan/softwares/my-jupyter-env/bin/pyccel", line 8, in <module>
    sys.exit(pyccel())
  File "/home/varadarajan/softwares/my-jupyter-env/lib/python3.10/site-packages/pyccel/commands/console.py", line 266, in pyccel
    execute_pyccel(filename,
  File "/home/varadarajan/softwares/my-jupyter-env/lib/python3.10/site-packages/pyccel/codegen/pipeline.py", line 378, in execute_pyccel
    src_compiler.compile_module(compile_obj=mod_obj,
  File "/home/varadarajan/softwares/my-jupyter-env/lib/python3.10/site-packages/pyccel/codegen/compiling/compilers.py", line 355, in compile_module
    self.run_command(cmd, verbose)
  File "/home/varadarajan/softwares/my-jupyter-env/lib/python3.10/site-packages/pyccel/codegen/compiling/compilers.py", line 500, in run_command
    raise RuntimeError(err_msg)
RuntimeError: Failed to build module
/home/varadarajan/softwares/Pyccel/my_progs/__pyccel__/test1_template.f90:162:8:

  162 |     c = a
      |        1
Warning: Extension: Conversion from INTEGER(4) to LOGICAL(1) at (1)
/home/varadarajan/softwares/Pyccel/my_progs/__pyccel__/test1_template.f90:146:8:

  146 |     c = a
      |        1
Warning: Extension: Conversion from INTEGER(1) to LOGICAL(1) at (1)
/home/varadarajan/softwares/Pyccel/my_progs/__pyccel__/test1_template.f90:114:8:

  114 |     c = a
      |        1
Warning: Extension: Conversion from INTEGER(8) to LOGICAL(1) at (1)
/home/varadarajan/softwares/Pyccel/my_progs/__pyccel__/test1_template.f90:98:8:

   98 |     c = a
      |        1
Warning: Extension: Conversion from INTEGER(2) to LOGICAL(1) at (1)
/home/varadarajan/softwares/Pyccel/my_progs/__pyccel__/test1_template.f90:82:8:

   82 |     c = a
      |        1
Error: Cannot convert REAL(4) to LOGICAL(1) at (1)
/home/varadarajan/softwares/Pyccel/my_progs/__pyccel__/test1_template.f90:66:8:

   66 |     c = a
      |        1
Error: Cannot convert COMPLEX(8) to LOGICAL(1) at (1)
/home/varadarajan/softwares/Pyccel/my_progs/__pyccel__/test1_template.f90:50:8:

   50 |     c = a
      |        1
Error: Cannot convert COMPLEX(4) to LOGICAL(1) at (1)
/home/varadarajan/softwares/Pyccel/my_progs/__pyccel__/test1_template.f90:34:8:

   34 |     c = a
      |        1
Error: Cannot convert REAL(8) to LOGICAL(1) at (1)
```

### Expected behavior
An error message why it cannot create a fortran executable. 

### Language
Fortran

### Additional context
If there is an bug id which tracks this issue, I can update the Pytest fortran skip part for #1762

### PR DESCRIPTION
**Pull request by EmilyBourne**
Add missing cast when creating an array from an object with a different datatype. Fixes #1785 

In order to call the cast on the argument passed to `np.array` (which may be a `InhomogeneousTupleType`) fix the type of the cast functions. 

Modify `InhomogeneousTupleType.datatype` to return a `FixedSizeType` if the datatypes of all elements are equivalent.

Also fix some minor bugs after #1756

### CODE REVIEW

**Review by smazouz42 — COMMENTED** [review]
looks good to me

**Comment by EmilyBourne** [comment]
> PR approved, looks good to me

@smazouz42 You don't seem to have approved. You have just commented. Please check the [docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews) to see how to actually approve a PR

**Review by smazouz42 — APPROVED** [review]
PR approved, looks good to me

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 1]
```diff
@@ -2175,10 +2175,20 @@ def create_float_array_tuple_ref(a : 'int[:,:]'):
         b = (a[0,:], a[1,:])
         c = array(b, dtype=float)
         return c
+    def create_bool_array_tuple_ref(a : 'int[:,:]'):
+        from numpy import array
+        b = (a[0,:], a[1,:])
+        c = array(b, dtype=float)
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:2181]: Should this not be an array of `bool`?
```suggestion
        c = array(b, dtype=bool)
```

File: pyccel/ast/builtins.py [hunk 2]
```diff
@@ -513,11 +513,6 @@ def __init__(self, *args):
 
         # Get possible datatypes
         dtypes = [a.class_type.datatype for a in args]
-        # Extract all dtypes inside any inhomogeneous tuples
-        while any(isinstance(d, InhomogeneousTupleType) for d in dtypes):
-            dtypes = [di for d in dtypes for di in ((d_elem.datatype for d_elem in d)
-                                                    if isinstance(d, InhomogeneousTupleType)
-                                                    else [d])]
```
Comment (yguclu) [inline pyccel/ast/builtins.py:520]: Why this change?

**Review by yguclu — CHANGES_REQUESTED** [review]
Looks good!

File: pyccel/ast/builtins.py (hunk 2 above)
Comment (EmilyBourne) [reply pyccel/ast/builtins.py:520]: This is now handled inside `InhomogeneousTupleType` as it is a reasonably common request. I am going to make a minor modification though to calculate this in the constructor instead of the `datatype` method

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 1 above)
Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: Oops you are right. The skip is also no longer necessary here. Fixed ff77793f6

**Review by yguclu — APPROVED** [review]

**Comment by yguclu** [comment]
@EmilyBourne Can you please mention the change to `datatype` in the PR description?

**Comment by EmilyBourne** [comment]
> @EmilyBourne Can you please mention the change to `datatype` in the PR description?

Done
